File: tools/crawl/crawl_tiktok.py
#!/usr/bin/env python3
"""
Crawl và tự động nhóm video TikTok theo Series / Phim.
Hỗ trợ lấy dữ liệu từ yt-dlp hoặc file JSON, gom nhóm qua Regex thông minh,
sắp xếp theo thứ tự phần (P1, P2, P3...), kiểm tra tính đầy đủ và lưu ra file/Supabase.

Cách dùng:
  python crawl_tiktok.py --channel https://www.tiktok.com/@username
  python crawl_tiktok.py --json channel.json --output series_output.json
  python crawl_tiktok.py --channel https://www.tiktok.com/@username --supabase
"""

import argparse
import json
import os
import re
import subprocess
import sys
import unicodedata
from datetime import datetime
from typing import Any, Dict, List, Optional, Tuple

# =====================================================================
# 1. Regex & Tiền xử lý Caption
# =====================================================================

def remove_accents(text: str) -> str:
    """Chuyển tiếng Việt có dấu sang không dấu để so khớp regex chuẩn xác."""
    text = unicodedata.normalize('NFD', text)
    text = re.sub(r'[\u0300-\u036f]', '', text)
    text = text.replace('đ', 'd').replace('Đ', 'D')
    return text.strip()

# Từ khóa đánh dấu tập cuối
FINAL_KEYWORDS = re.compile(
    r'\b(?:part\s*cuoi|phan\s*cuoi|tap\s*cuoi|ky\s*cuoi|final(?:\s*part)?|finale|the\s*end|ending|ket\s*thuc|het|full\s*review|end)\b',
    re.IGNORECASE
)

# Các mẫu nhận diện số tập: "P1", "Phần 2", "Tập 3", "Ep 04", "Part 5", "Tập 3/5", "#02"
PART_PATTERNS = [
    # Mẫu 1: "phần 2", "tap 3", "p 01", "ep 4", "part 5" kèm "/10" nếu có
    re.compile(r'\b(?:part|phan|tap|ep|episode|p)\s*[.:\-_]?\s*0*(\d{1,3})\b(?:\s*(?:\/|of|tren)\s*0*(\d{1,3}))?', re.IGNORECASE),
    # Mẫu 2: "#03" hoặc "#3/5"
    re.compile(r'#\s*0*(\d{1,3})\b(?:\s*\/\s*0*(\d{1,3}))?', re.IGNORECASE),
    # Mẫu 3: "3/5" hoặc "3 of 5"
    re.compile(r'\b0*(\d{1,3})\s*(?:\/|of)\s*0*(\d{1,3})\b', re.IGNORECASE),
]

# Từ khóa nhiễu cần lọc khi rút tên phim
NOISE_KEYWORDS = re.compile(
    r'\b(?:review|tom\s*tat|full|hd|4k|vietsub|thuyet\s*minh|phim|movie|series|official|trailer|reaction|spoiler|toan\s*bo|hay\s*nhat|kenh|tiktok)\b',
    re.IGNORECASE
)

def extract_part_info(raw_title: str) -> Dict[str, Any]:
    """Nhận diện số tập/phần từ caption TikTok."""
    norm = remove_accents(raw_title)
    is_final = bool(FINAL_KEYWORDS.search(norm))
    
    for pattern in PART_PATTERNS:
        match = pattern.search(norm)
        if match:
            part_num = int(match.group(1))
            total_parts = int(match.group(2)) if match.group(2) else None
            # Tránh nhầm ngày tháng như 12/5 (nếu tổng < số tập hiện tại thì bỏ qua)
            if total_parts is not None and total_parts < part_num:
                total_parts = None
            return {
                'part_number': part_num,
                'total_parts': total_parts,
                'is_final': is_final,
                'confidence': 0.9
            }
            
    return {
        'part_number': None,
        'total_parts': None,
        'is_final': is_final,
        'confidence': 0.4 if is_final else 0.0
    }

def extract_series_name(raw_title: str) -> str:
    """
    Rút trích tên bộ phim/series sạch sẽ:
    - Loại bỏ hashtag (#fyp, #xuhuong, #reviewphim)
    - Loại bỏ đánh số phần (P1, Tập 2, Part 3)
    - Loại bỏ các từ quảng cáo rác (review, tóm tắt phim...)
    - Giữ lại tên phim có nghĩa
    """
    # 1. Bỏ hashtag
    text = re.sub(r'#\S+', ' ', raw_title)
    
    # 2. Bỏ các mẫu số phần
    text = re.sub(r'\b(?:part|phan|phần|tap|tập|ep|episode|p)\s*[.:\-_]?\s*0*(\d{1,3})\b(?:\s*(?:\/|of|trên|tren)\s*0*(\d{1,3}))?', ' ', text, flags=re.IGNORECASE)
    text = re.sub(r'#\s*0*(\d{1,3})\b(?:\s*\/\s*0*(\d{1,3}))?', ' ', text)
    text = re.sub(r'\b0*(\d{1,3})\s*(?:\/|of)\s*0*(\d{1,3})\b', ' ', text)
    text = FINAL_KEYWORDS.sub(' ', text)
    
    # 3. Bỏ từ khóa thừa
    text = re.sub(r'\b(?:review|tóm\s*tắt|tom\s*tat|full|hd|4k|vietsub|thuyết\s*minh|thuyet\s*minh|phim|movie|series|official|trailer|reaction|spoiler|toàn\s*bộ|hay\s*nhất)\b', ' ', text, flags=re.IGNORECASE)
    
    # 4. Làm sạch ký tự ngăn cách
    text = re.sub(r'[|\-_–—:;,.!?]+', ' ', text)
    text = re.sub(r'\s+', ' ', text).strip()
    
    # Nếu bị xóa hết thì trả về 30 ký tự đầu của caption gốc
    return text if len(text) >= 2 else raw_title[:40].strip()

def normalize_series_key(name: str) -> str:
    """Tạo key duy nhất không dấu để gom các video cùng một phim về một nhóm."""
    clean = remove_accents(name).lower()
    clean = re.sub(r'[^a-z0-9]+', ' ', clean).strip()
    return clean

# =====================================================================
# 2. Cào video TikTok bằng yt-dlp
# =====================================================================

def fetch_tiktok_channel_videos(channel_url: str) -> Tuple[List[Dict[str, Any]], Dict[str, Any]]:
    """
    Sử dụng yt-dlp để lấy toàn bộ danh sách video công khai của kênh TikTok.
    Lệnh: yt-dlp --flat-playlist -J "<url>"
    """
    print(f"🚀 Đang quét video từ kênh: {channel_url} ...")
    cmd = [
        'yt-dlp',
        '--flat-playlist',
        '-J',
        '--extractor-args', 'tiktok:api_hostname=api22-normal-c-useast1a.tiktokv.com',
        channel_url
    ]
    
    try:
        proc = subprocess.run(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True, check=True)
        data = json.loads(proc.stdout)
    except FileNotFoundError:
        print("❌ Lỗi: Chưa cài đặt 'yt-dlp'. Vui lòng cài đặt: pip install yt-dlp hoặc choco install yt-dlp")
        sys.exit(1)
    except subprocess.CalledProcessError as e:
        print(f"❌ Lỗi khi chạy yt-dlp: {e.stderr}")
        sys.exit(1)
    except json.JSONDecodeError as e:
        print(f"❌ Lỗi parse JSON từ yt-dlp: {e}")
        sys.exit(1)

    entries = data.get('entries', []) or []
    creator_info = {
        'creator_id': data.get('channel_id') or data.get('uploader_id') or data.get('uploader') or 'tiktok_creator',
        'creator_name': data.get('channel') or data.get('uploader') or channel_url.split('@')[-1].split('/')[0],
        'creator_url': channel_url
    }
    
    print(f"✅ Đã tìm thấy {len(entries)} video từ @{creator_info['creator_name']}")
    return entries, creator_info

# =====================================================================
# 3. Pipeline Gom nhóm & Sắp xếp Series
# =====================================================================

def process_and_group_videos(entries: List[Dict[str, Any]], creator_info: Dict[str, Any]) -> List[Dict[str, Any]]:
    """
    Xử lý danh sách video:
    1. Parse từng caption -> Part Info & Clean Title
    2. Gom nhóm theo Series Key
    3. Sắp xếp thứ tự các phần từ 1 -> N
    4. Đánh giá tính hoàn thành của series
    """
    series_map: Dict[str, Dict[str, Any]] = {}
    
    for item in entries:
        raw_title = item.get('title') or item.get('description') or ''
        video_id = str(item.get('id') or '')
        url = item.get('url') or item.get('webpage_url') or f"https://www.tiktok.com/@{creator_info['creator_name']}/video/{video_id}"
        thumbnail = item.get('thumbnail') or (item.get('thumbnails', [{}])[-1].get('url') if item.get('thumbnails') else None)
        timestamp = item.get('timestamp') or item.get('upload_date')
        published_at = datetime.fromtimestamp(timestamp).isoformat() if isinstance(timestamp, (int, float)) else str(timestamp)
        duration = item.get('duration')
        
        part_info = extract_part_info(raw_title)
        display_title = extract_series_name(raw_title)
        series_key = normalize_series_key(display_title)
        
        if not series_key:
            series_key = "video_le"
            display_title = "Video đơn lẻ"
            
        video_obj = {
            'video_id': video_id,
            'title': raw_title,
            'clean_title': display_title,
            'url': url,
            'embed_url': f"https://www.tiktok.com/embed/v2/{video_id}",
            'thumbnail': thumbnail,
            'duration': duration,
            'published_at': published_at,
            'part_number': part_info['part_number'],
            'total_parts': part_info['total_parts'],
            'is_final': part_info['is_final'],
        }
        
        if series_key not in series_map:
            series_map[series_key] = {
                'series_key': f"tiktok:{creator_info['creator_id']}:{series_key}",
                'title': display_title,
                'creator_id': creator_info['creator_id'],
                'creator_name': creator_info['creator_name'],
                'creator_url': creator_info['creator_url'],
                'cover': thumbnail,
                'videos': []
            }
            
        series_map[series_key]['videos'].append(video_obj)

    # Sắp xếp các tập trong mỗi series
    result_series = []
    for s_key, s_data in series_map.items():
        videos = s_data['videos']
        
        # Sắp xếp: ưu tiên số tập (1, 2, 3...), nếu không có số tập thì theo ngày đăng cũ -> mới
        def sort_key(v):
            p = v['part_number'] if v['part_number'] is not None else 9999
            return (p, v.get('published_at') or '')
            
        sorted_videos = sorted(videos, key=sort_key)
        
        # Đánh giá hoàn thành (Completion Check)
        part_numbers = [v['part_number'] for v in sorted_videos if v['part_number'] is not None]
        has_final = any(v['is_final'] for v in sorted_videos)
        
        status = 'UNKNOWN'
        if len(part_numbers) > 1:
            is_sequential = part_numbers == list(range(1, len(part_numbers) + 1))
            if is_sequential and has_final:
                status = 'COMPLETE'
            elif is_sequential:
                status = 'IN_PROGRESS'
            else:
                status = 'INCOMPLETE'  # Thiếu tập ở giữa
        elif len(sorted_videos) == 1 and not has_final and not part_numbers:
            status = 'SINGLE'
            
        s_data['videos'] = sorted_videos
        s_data['video_count'] = len(sorted_videos)
        s_data['status'] = status
        s_data['found_parts'] = len(part_numbers)
        result_series.append(s_data)

    # Sắp xếp series nhiều video lên đầu
    result_series.sort(key=lambda s: s['video_count'], reverse=True)
    return result_series

# =====================================================================
# 4. Lưu vào Supabase (Tùy chọn)
# =====================================================================

def save_to_supabase(series_list: List[Dict[str, Any]], creator_info: Dict[str, Any]):
    """Upsert kết quả cào vào Supabase Database."""
    from dotenv import load_dotenv
    load_dotenv()
    
    supabase_url = os.getenv('VITE_SUPABASE_URL')
    service_key = os.getenv('SUPABASE_SERVICE_ROLE_KEY')
    
    if not supabase_url or not service_key:
        print("⚠️ Thiếu VITE_SUPABASE_URL hoặc SUPABASE_SERVICE_ROLE_KEY trong .env. Bỏ qua bước lưu Supabase.")
        return

    try:
        from supabase import create_client
        supabase = create_client(supabase_url, service_key)
        
        # 1. Upsert Creator
        supabase.table('review_creators').upsert({
            'platform': 'tiktok',
            'creator_url': creator_info['creator_url'],
            'creator_id': str(creator_info['creator_id']),
            'creator_name': creator_info['creator_name'],
            'last_synced_at': datetime.utcnow().isoformat()
        }, on_conflict='platform,creator_url').execute()
        
        # 2. Upsert Series & Videos
        for s in series_list:
            supabase.table('review_series').upsert({
                'series_key': s['series_key'],
                'platform': 'tiktok',
                'creator_id': str(s['creator_id']),
                'creator_name': s['creator_name'],
                'title': s['title'],
                'movie_title': s['title'],
                'status': 'COMPLETE' if s['status'] == 'COMPLETE' else 'UNKNOWN',
                'found_parts': s['found_parts'],
                'updated_at': datetime.utcnow().isoformat()
            }, on_conflict='series_key').execute()
            
            for v in s['videos']:
                supabase.table('review_videos').upsert({
                    'platform': 'tiktok',
                    'video_id': v['video_id'],
                    'series_key': s['series_key'],
                    'creator_id': str(s['creator_id']),
                    'creator_name': s['creator_name'],
                    'title': v['title'],
                    'canonical_url': v['url'],
                    'embed_url': v['embed_url'],
                    'thumbnail': v['thumbnail'],
                    'duration': v['duration'],
                    'part_number': v['part_number'],
                    'total_parts': v['total_parts'],
                    'is_final': v['is_final'],
                    'last_seen_at': datetime.utcnow().isoformat()
                }, on_conflict='platform,video_id').execute()
                
        print("🎉 Đã lưu toàn bộ series và video vào cơ sở dữ liệu Supabase thành công!")
    except Exception as e:
        print(f"❌ Lỗi khi lưu vào Supabase: {e}")

# =====================================================================
# 5. CLI Main Execution
# =====================================================================

def main():
    parser = argparse.ArgumentParser(description="Pipeline cào & nhóm series TikTok tự động.")
    parser.add_argument('--channel', type=str, help="Link kênh TikTok (vd: https://www.tiktok.com/@reviewphimhay)")
    parser.add_argument('--json', type=str, help="Đọc dữ liệu từ file JSON có sẵn")
    parser.add_argument('--output', type=str, default="tiktok_series.json", help="File output kết quả JSON")
    parser.add_argument('--supabase', action='store_true', help="Tự động đồng bộ lên Supabase Database")
    
    args = parser.parse_args()
    
    if not args.channel and not args.json:
        print("⚠️ Vui lòng cung cấp --channel <url> hoặc --json <path_to_json>")
        parser.print_help()
        sys.exit(1)
        
    entries = []
    creator_info = {'creator_id': 'unknown', 'creator_name': 'TikTok Creator', 'creator_url': ''}
    
    if args.json:
        with open(args.json, 'r', encoding='utf-8') as f:
            raw_data = json.load(f)
            if isinstance(raw_data, dict):
                entries = raw_data.get('entries', [raw_data])
                creator_info['creator_name'] = raw_data.get('uploader') or 'tiktok_user'
                creator_info['creator_id'] = raw_data.get('uploader_id') or creator_info['creator_name']
            elif isinstance(raw_data, list):
                entries = raw_data
        print(f"📂 Đã nạp {len(entries)} video từ file JSON: {args.json}")
    elif args.channel:
        entries, creator_info = fetch_tiktok_channel_videos(args.channel)
        
    series_list = process_and_group_videos(entries, creator_info)
    
    print("\n" + "=" * 60)
    print(f"📊 KẾT QUẢ GOM SERIES: @{creator_info['creator_name']}")
    print(f"Tổng số series tìm thấy: {len(series_list)}")
    print("=" * 60)
    
    for i, s in enumerate(series_list[:10], 1):
        print(f"\n🎬 {i}. {s['title']} ({s['video_count']} video, Trạng thái: {s['status']})")
        for v in s['videos'][:5]:
            p_str = f"P{v['part_number']}" if v['part_number'] else "Tập ?"
            print(f"   - {p_str}: {v['url']}")
        if len(s['videos']) > 5:
            print(f"   ... và {len(s['videos']) - 5} video khác")
            
    # Lưu ra file JSON
    with open(args.output, 'w', encoding='utf-8') as f:
        json.dump({
            'creator': creator_info,
            'total_series': len(series_list),
            'total_videos': len(entries),
            'series': series_list
        }, f, ensure_ascii=False, indent=2)
    print(f"\n💾 Đã lưu kết quả hoàn chỉnh vào: {args.output}")
    
    if args.supabase:
        save_to_supabase(series_list, creator_info)

if __name__ == '__main__':
    main()

File: tools/crawl/crawl_tiktok.mjs
#!/usr/bin/env node
/**
 * Pipeline cào & nhóm Series TikTok tự động bằng Node.js.
 * Hỗ trợ chạy trực tiếp với yt-dlp, đọc file JSON hoặc dán link.
 *
 * Cách dùng:
 *   node crawl_tiktok.mjs --channel https://www.tiktok.com/@username
 *   node crawl_tiktok.mjs --json channel.json --output series.json
 *   node crawl_tiktok.mjs --channel https://www.tiktok.com/@username --supabase
 */

import { config } from 'dotenv'
import { createClient } from '@supabase/supabase-js'
import { execFile } from 'child_process'
import { promisify } from 'util'
import fs from 'fs/promises'
import path from 'path'

config()
const execFileAsync = promisify(execFile)

const supabaseUrl = process.env.VITE_SUPABASE_URL
const supabaseServiceKey = process.env.SUPABASE_SERVICE_ROLE_KEY
const supabase = (supabaseUrl && supabaseServiceKey)
  ? createClient(supabaseUrl, supabaseServiceKey, { auth: { persistSession: false } })
  : null

function removeAccents(str = '') {
  return str
    .normalize('NFD')
    .replace(/[\u0300-\u036f]/g, '')
    .replace(/[đĐ]/g, 'd')
    .toLowerCase()
    .trim()
}

const FINAL_REGEX =
  /\b(?:part\s*(?:cuối|cuoi)|(?:phần|phan)\s*(?:cuối|cuoi)|(?:tập|tap)\s*(?:cuối|cuoi)|(?:kỳ|ky)\s*(?:cuối|cuoi)|final(?:\s*part)?|finale|the\s*end|ending|(?:kết\s*thúc|ket\s*thuc)|hết|het|full\s*review|end)\b/i
const PART_PATTERNS = [
  /\b(?:part|phan|tap|ep|episode|p)\s*[.:\-_]?\s*0*(\d{1,3})\b(?:\s*(?:\/|of|tren)\s*0*(\d{1,3}))?/i,
  /#\s*0*(\d{1,3})\b(?:\s*\/\s*0*(\d{1,3}))?/i,
  /\b0*(\d{1,3})\s*(?:\/|of)\s*0*(\d{1,3})\b/i,
]

export function extractPartInfo(title = '') {
  const norm = removeAccents(title)
  const isFinal = FINAL_REGEX.test(norm)

  for (const pattern of PART_PATTERNS) {
    const match = norm.match(pattern)
    if (match) {
      const partNum = parseInt(match[1], 10)
      let totalParts = match[2] ? parseInt(match[2], 10) : null
      if (totalParts !== null && totalParts < partNum) totalParts = null
      return { partNumber: partNum, totalParts, isFinal, confidence: 0.9 }
    }
  }

  return { partNumber: null, totalParts: null, isFinal, confidence: isFinal ? 0.4 : 0 }
}

export function extractSeriesName(title = '') {
  let text = title.replace(/#\S+/g, ' ')
  text = text.replace(/\b(?:part|phan|phần|tap|tập|ep|episode|p)\s*[.:\-_]?\s*0*(\d{1,3})\b(?:\s*(?:\/|of|trên|tren)\s*0*(\d{1,3}))?/gi, ' ')
  text = text.replace(/#\s*0*(\d{1,3})\b(?:\s*\/\s*0*(\d{1,3}))?/gi, ' ')
  text = text.replace(/\b0*(\d{1,3})\s*(?:\/|of)\s*0*(\d{1,3})\b/gi, ' ')
  text = text.replace(FINAL_REGEX, ' ')
  text = text.replace(/\b(?:review|tóm\s*tắt|tom\s*tat|full|hd|4k|vietsub|thuyết\s*minh|thuyet\s*minh|phim|movie|series|official|trailer|reaction|spoiler|toàn\s*bộ|hay\s*nhất)\b/gi, ' ')
  text = text.replace(/[|\-_–—:;,.!?]+/g, ' ').replace(/\s+/g, ' ').trim()
  return text.length >= 2 ? text : title.slice(0, 40).trim()
}

export function normalizeSeriesKey(name = '') {
  return removeAccents(name).replace(/[^a-z0-9]+/g, ' ').trim()
}

export function groupVideosIntoSeries(entries, creatorInfo) {
  const seriesMap = new Map()

  for (const item of entries) {
    const rawTitle = item.title || item.description || ''
    const videoId = String(item.id || '')
    const url = item.url || item.webpage_url || `https://www.tiktok.com/@${creatorInfo.creator_name}/video/${videoId}`
    const thumbnail = item.thumbnail || (Array.isArray(item.thumbnails) ? item.thumbnails[item.thumbnails.length - 1]?.url : null)
    const timestamp = item.timestamp || item.upload_date
    const publishedAt = typeof timestamp === 'number' ? new Date(timestamp * 1000).toISOString() : String(timestamp || new Date().toISOString())
    const duration = item.duration || null

    const partInfo = extractPartInfo(rawTitle)
    const displayTitle = extractSeriesName(rawTitle)
    let seriesKey = normalizeSeriesKey(displayTitle)
    if (!seriesKey) seriesKey = 'video_le'

    const videoObj = {
      video_id: videoId,
      title: rawTitle,
      clean_title: displayTitle,
      url,
      embed_url: `https://www.tiktok.com/embed/v2/${videoId}`,
      thumbnail,
      duration,
      published_at: publishedAt,
      part_number: partInfo.partNumber,
      total_parts: partInfo.totalParts,
      is_final: partInfo.isFinal,
    }

    if (!seriesMap.has(seriesKey)) {
      seriesMap.set(seriesKey, {
        series_key: `tiktok:${creatorInfo.creator_id}:${seriesKey}`,
        title: displayTitle,
        creator_id: creatorInfo.creator_id,
        creator_name: creatorInfo.creator_name,
        creator_url: creatorInfo.creator_url,
        cover: thumbnail,
        videos: [],
      })
    }

    seriesMap.get(seriesKey).videos.push(videoObj)
  }

  const resultSeries = []
  for (const [, sData] of seriesMap) {
    const sorted = [...sData.videos].sort((a, b) => {
      const pa = a.part_number !== null ? a.part_number : 9999
      const pb = b.part_number !== null ? b.part_number : 9999
      if (pa !== pb) return pa - pb
      return (a.published_at || '').localeCompare(b.published_at || '')
    })

    const parts = sorted.map((v) => v.part_number).filter((p) => p !== null)
    const hasFinal = sorted.some((v) => v.is_final)
    let status = 'UNKNOWN'
    if (parts.length > 1) {
      const isSeq = parts.every((p, idx) => p === idx + 1)
      if (isSeq && hasFinal) status = 'COMPLETE'
      else if (isSeq) status = 'IN_PROGRESS'
      else status = 'INCOMPLETE'
    } else if (sorted.length === 1 && !hasFinal && parts.length === 0) {
      status = 'SINGLE'
    }

    resultSeries.push({
      ...sData,
      videos: sorted,
      video_count: sorted.length,
      status,
      found_parts: parts.length,
    })
  }

  resultSeries.sort((a, b) => b.video_count - a.video_count)
  return resultSeries
}

async function fetchFromYtDlp(channelUrl) {
  console.log(`🚀 Đang quét video từ kênh: ${channelUrl} ...`)
  const args = [
    '--flat-playlist',
    '-J',
    '--extractor-args',
    'tiktok:api_hostname=api22-normal-c-useast1a.tiktokv.com',
    channelUrl,
  ]

  const { stdout } = await execFileAsync('yt-dlp', args, { maxBuffer: 100 * 1024 * 1024 })
  const data = JSON.parse(stdout)
  const entries = data.entries || []
  const creatorInfo = {
    creator_id: data.channel_id || data.uploader_id || data.uploader || 'tiktok_creator',
    creator_name: data.channel || data.uploader || channelUrl.split('@')[1]?.split('/')[0] || 'tiktok_user',
    creator_url: channelUrl,
  }

  console.log(`✅ Đã tìm thấy ${entries.length} video từ @${creatorInfo.creator_name}`)
  return { entries, creatorInfo }
}

async function saveToSupabase(seriesList, creatorInfo) {
  if (!supabase) {
    console.log('⚠️ Thiếu VITE_SUPABASE_URL hoặc SUPABASE_SERVICE_ROLE_KEY trong .env. Bỏ qua Supabase.')
    return
  }

  console.log('💾 Đang lưu dữ liệu vào Supabase...')
  await supabase.from('review_creators').upsert({
    platform: 'tiktok',
    creator_url: creatorInfo.creator_url,
    creator_id: String(creatorInfo.creator_id),
    creator_name: creatorInfo.creator_name,
    last_synced_at: new Date().toISOString(),
  }, { onConflict: 'platform,creator_url' })

  for (const s of seriesList) {
    await supabase.from('review_series').upsert({
      series_key: s.series_key,
      platform: 'tiktok',
      creator_id: String(s.creator_id),
      creator_name: s.creator_name,
      title: s.title,
      movie_title: s.title,
      status: s.status === 'COMPLETE' ? 'COMPLETE' : 'UNKNOWN',
      found_parts: s.found_parts,
      updated_at: new Date().toISOString(),
    }, { onConflict: 'series_key' })

    for (const v of s.videos) {
      await supabase.from('review_videos').upsert({
        platform: 'tiktok',
        video_id: v.video_id,
        series_key: s.series_key,
        creator_id: String(s.creator_id),
        creator_name: s.creator_name,
        title: v.title,
        canonical_url: v.url,
        embed_url: v.embed_url,
        thumbnail: v.thumbnail,
        duration: v.duration,
        part_number: v.part_number,
        total_parts: v.total_parts,
        is_final: v.is_final,
        last_seen_at: new Date().toISOString(),
      }, { onConflict: 'platform,video_id' })
    }
  }

  console.log('🎉 Đã đồng bộ toàn bộ Series và Video TikTok vào Supabase thành công!')
}

async function main() {
  const args = process.argv.slice(2)
  let channelUrl = ''
  let jsonPath = ''
  let outputPath = 'tiktok_series.json'
  let shouldSaveSupabase = false

  for (let i = 0; i < args.length; i++) {
    if (args[i] === '--channel' && args[i + 1]) channelUrl = args[++i]
    else if (args[i] === '--json' && args[i + 1]) jsonPath = args[++i]
    else if (args[i] === '--output' && args[i + 1]) outputPath = args[++i]
    else if (args[i] === '--supabase') shouldSaveSupabase = true
  }

  if (!channelUrl && !jsonPath) {
    console.log(`
Cách dùng:
  node crawl_tiktok.mjs --channel https://www.tiktok.com/@username
  node crawl_tiktok.mjs --json channel.json --output result.json
  node crawl_tiktok.mjs --channel https://www.tiktok.com/@username --supabase
    `)
    return
  }

  let entries = []
  let creatorInfo = { creator_id: 'unknown', creator_name: 'TikTok Creator', creator_url: channelUrl }

  if (jsonPath) {
    const raw = await fs.readFile(path.resolve(jsonPath), 'utf-8')
    const parsed = JSON.parse(raw)
    entries = Array.isArray(parsed) ? parsed : (parsed.entries || [parsed])
    creatorInfo.creator_name = parsed.uploader || parsed.channel || 'tiktok_user'
    creatorInfo.creator_id = parsed.uploader_id || creatorInfo.creator_name
    console.log(`📂 Đã nạp ${entries.length} video từ file JSON: ${jsonPath}`)
  } else if (channelUrl) {
    const res = await fetchFromYtDlp(channelUrl)
    entries = res.entries
    creatorInfo = res.creatorInfo
  }

  const seriesList = groupVideosIntoSeries(entries, creatorInfo)

  console.log('\n============================================================')
  console.log(`📊 KẾT QUẢ GOM SERIES: @${creatorInfo.creator_name}`)
  console.log(`Tổng số series tìm thấy: ${seriesList.length}`)
  console.log('============================================================')

  for (let i = 0; i < Math.min(10, seriesList.length); i++) {
    const s = seriesList[i]
    console.log(`\n🎬 ${i + 1}. ${s.title} (${s.video_count} video, Trạng thái: ${s.status})`)
    for (const v of s.videos.slice(0, 4)) {
      const p = v.part_number ? `P${v.part_number}` : 'Tập ?'
      console.log(`   - ${p}: ${v.url}`)
    }
  }

  await fs.writeFile(outputPath, JSON.stringify({
    creator: creatorInfo,
    total_series: seriesList.length,
    total_videos: entries.length,
    series: seriesList,
  }, null, 2), 'utf-8')
  console.log(`\n💾 Đã lưu kết quả vào: ${outputPath}`)

  if (shouldSaveSupabase) {
    await saveToSupabase(seriesList, creatorInfo)
  }
}

// Chạy trực tiếp qua CLI
const isDirectExecution = process.argv[1] && (
  process.argv[1].endsWith('crawl_tiktok.mjs') ||
  process.argv[1].endsWith('crawl_tiktok.js')
)

if (isDirectExecution) {
  main().catch((err) => {
    console.error('❌ Lỗi:', err)
    process.exit(1)
  })
}
